### PR TITLE
13 pawn pathfinding

### LINKED ISSUE #13: Pawn Pathfinding
## Description
As a player I want my pawns to intelligently move to the places they need to because I don't want to have to wait.

### Functional
- [x] Pawns no longer randomly pick destinations in the world
- [x] Pawns know where they should go
- [x] Pawns can generate a path to their destination
- [x] Pawns follow the path
- [x] Obstacle avoidance

## Dependencies
- #9

### PR DESCRIPTION
**Pull request by martinjonsson01**
## Summary
Adds pawn pathfinding using the A* algorithm. Changes pawns to randomly decide spots to pathfind to.

xMDp1QH4GI

## Test Coverage
image
image

## Definition of Done
- [x] Completed all Tasks of the User Story
- [x] The Tasks have tests and they all pass
- [x] Overall test coverage is >= 90%
- [x] A UML class diagram of the newly implemented User Story is done and included in the README.md of the package

### com.thebois.models.beings.pathfinding
com thebois models beings pathfinding
### com.thebois.listeners.events
com thebois listeners events
### com.thebois.models.beings
com thebois models beings
### com.thebois.models.world
com thebois models world

Closes #13

### CODE REVIEW

File: src/main/java/com/thebois/views/IView.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/IView.java:16]: "."

File: src/test/java/com/thebois/testutils/MockUtils.java
Comment (jonathan-lindqvist) [inline src/test/java/com/thebois/testutils/MockUtils.java:9]: Huh?

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Nice 👍 

Small fixes and explanations then it will be good to go!

I will keep pushing smaller PRs, but this was MUCH better 😉

File: src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java:24]: Will clarify the comment.

File: src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java:53]: "This" as in the `costComparator`? The comparator is used in the `PriorityQueue<ITile> openSet` just below, or did you mean something else?

File: src/main/java/com/thebois/models/beings/pathfinding/TileCostComparator.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/pathfinding/TileCostComparator.java:24]: Thing is, comparators are a specific construct in Java. They are used by collections like the `PriorityQueue` (used in the `AstarPathFinder`) to sort elements. It has to be in a separate class, because it's implementing `Comparator<ITile>`. An anonymous inner class would be even worse, wouldn't it?

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java:42]: As a *default*, yeah. Maybe I could set it at 1, but that's also arbitrary. 

Though, seeing as this is in *Abstract*Terrain, I could leave it unimplemented and force all subtypes to implement it themselves.

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java:47]: They're for debugging, the integrated IDEA debugger and test runner actually calls these methods when rendering debug info, so it's really useful!

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:37]: Yeah... Did you see the line right below it? lol

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:66]: Yeah, repopulating it from being empty. 

It was called just populate before I had to call it when new structures are placed. Calling `populateCanonicalMatrix` every time a new structure is placed seemed even worse than calling `repopulateCanonicalMatrix` once in the constructor, no?

File: src/test/java/com/thebois/testutils/MockUtils.java
Comment (martinjonsson01) [reply src/test/java/com/thebois/testutils/MockUtils.java:9]: ... I was supposed to move some things into here...


...


I never did.

File: src/main/java/com/thebois/models/beings/pathfinding/TileCostComparator.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/pathfinding/TileCostComparator.java:24]: Maybe you missed the comment:
```java
// The priority queue automatically sorts elements according to the comparator.
this.openSet = new PriorityQueue<>(costComparator);
```

File: src/main/java/com/thebois/models/world/Grass.java
Comment (Matte02) [inline src/main/java/com/thebois/models/world/Grass.java:42]: How wrong would it be to move this up to `Abstract Terrain`, and let specific terrrain overried this if it should be something else?

File: src/main/java/com/thebois/models/world/structures/House.java
Comment (Matte02) [inline src/main/java/com/thebois/models/world/structures/House.java:37]: Same as terrain, wouldn't it be better to move this up to `Abstract Structre` since this should be the same for most structures.

**Review by Matte02 — CHANGES_REQUESTED** [review]
Well done! 

Didn't find anything major. Only some small style things and a few questions.

File: src/main/java/com/thebois/listeners/events/ObstaclePlacedEvent.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/listeners/events/ObstaclePlacedEvent.java:11]: `Position` uses floats, and using it would mean having to cast the values to `int` at multiple places. This way, only valid obstacle positions make it into the event. The getter `getPosition` still returns a `Position` to make using the event easier.

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:174]: That's entirely possible, but it'll get hard to manage in the future really quickly as we add more and more dynamic obstacles. If you have to manually add the new obstacles to the matrix every time you create a new one, then if you miss a spot you won't know why some obstacles aren't being included. By completely recalculating the entire matrix you make sure that it contains everything it should, and won't miss things.

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (Matte02) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java]: Wait, just realized I said something completely different later. IMO this should be Float.MAX as it should be the base case for all structures.

File: src/main/java/com/thebois/models/world/Grass.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/Grass.java:42]: 😂 it was that way at first but @jonathan-lindqvist didn't like that so I refactored it like this.

The benefit of doing it the way you and I thought is that you don't have to implement it in every subtype if they all return 0 anyways. 

The downside to that is when you forget to override the method in a subtype that actually has a bigger cost, you won't know until you run the game and try to get a pawn to pathfind over the tile to see if it avoids it or not. 

So the question is, do we think we'll have more tiles with 0 cost, or more with specified costs?

File: src/main/java/com/thebois/models/world/World.java
Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:174]: Fair enough, and if we at any point run into performance issues then we can go back and have a look at this to see if we can optimize it.

File: src/main/java/com/thebois/models/world/structures/House.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/structures/House.java:37]: Either way we go, it should be the same for all `ITile` implementations.

File: src/main/java/com/thebois/models/world/Grass.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/Grass.java:42]: On second thought, you're probably right that most, if not all, terrain will have a movement cost of 0. Then we should probably keep it with a default value in the abstract class and have types override it when needed.

Same goes for structures, most, if not all, will not be walkable at all.

When implementing resources, they should all probably either be unwalkable or very slowly walkable. 

@jonathan-lindqvist thoughts?

Comment (Matte02) [reply src/main/java/com/thebois/models/world/Grass.java:42]: I agree with this.

File: src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/pathfinding/AstarPathFinder.java:61]: Nah missed the line, probably read it as costCalculator which made it all very confusing.

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java]: Either is fine, talked a bit with @Matte02 about this, if all terrain like sand/dirt/grass has the same speed, sure keep it in abstract but if we want dirt/sand to have a slower speed (imo they should) then leave it to the classes to implement it.

File: src/main/java/com/thebois/models/world/World.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/World.java:37]: But why do we need a constructor that takes a colony? We will never have a colony before the world. If this code is only the second constructor is made only for tests isn't that a code smell?

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java]: OK, then we should leave it as-is with the abstract class forcing subtypes to implement it themselves. 

I'll refactor structures to work the same, then.

File: src/main/java/com/thebois/models/world/World.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/World.java:66]: if you repopulate it you also populate it, correct?

Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/World.java:66]: Sounds a bit weird to populate every time a structure is placed no?

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Continue discussion items.

File: src/main/java/com/thebois/models/world/AbstractTerrain.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/AbstractTerrain.java]: 👍

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:37]: Don't know, I didn't make that constructor :P

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:37]: IIRC @Matte02 has already removed that constructor in his branch.

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:66]: >Sounds a bit weird to populate every time a structure is placed no?

You're contradicting yourself, but yeah I agree with this, that's why I called it `repopulate` since it's repopulating the matrix every time a structure is placed.

I'll just rename it to `calculateCanonicalMatrix`.

**Review by JacobBredin — APPROVED** [review]
Lgtm! Great work!

**Comment by martinjonsson01** [comment]
>wrong PR sorry!

Damn, does that mean I haven't done great work? :(((

File: src/main/java/com/thebois/models/Position.java
Comment (JacobBredin) [inline src/main/java/com/thebois/models/Position.java:48]: Being punctual I see 👍

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/Position.java:48]: Btw you should accept/request changes instead of commenting, to actually affect the status of the PR!

Comment (JacobBredin) [reply src/main/java/com/thebois/models/Position.java:48]: Thanks for the pro-tip

File: src/main/java/com/thebois/models/Position.java
Comment (JacobBredin) [inline src/main/java/com/thebois/models/Position.java:119]: I know everything is done with whole numbers, but the return value should probably be rounded

**Review by JacobBredin — CHANGES_REQUESTED** [review]

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (JacobBredin) [inline src/main/java/com/thebois/models/beings/AbstractBeing.java:19]: Question

If the path towards the Beings destination gets interrupted by something, how does the Being remember what destination it was supposed to go to if this variable is removed?

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:19]: It knows its previous path, so when it recalculates its new path, it uses the destination of the previous path. See the `onObstaclePlaced` override method on lin 52!

File: src/main/java/com/thebois/models/world/IWorld.java
Comment (JacobBredin) [inline src/main/java/com/thebois/models/world/IWorld.java:8]: I hope we can look into merging this functionality with IFinder later. I think it would fit perfectly, some functionality is already implemented (in my branch)

**Review by JacobBredin — CHANGES_REQUESTED** [review]

File: src/main/java/com/thebois/models/world/IWorld.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/IWorld.java:8]: Yup, that's probably a good idea!

(this is a comment, not a request for change :P )

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/IWorld.java:8]: Also, you can start your review, leave all your comments/requests for changes, and then submit them all at once, instead of reviewing multiple times!

Comment (JacobBredin) [reply src/main/java/com/thebois/models/world/IWorld.java:8]: But I love making multiple reviews! Also it is "important Feedback you must address before merging" 😉

**Review by Matte02 — APPROVED** [review]
Great work!

**Review by JacobBredin — APPROVED** [review]
LGTM! Great Work!

**Review by jonathan-lindqvist — APPROVED** [review]
Noice

**Pull request merged**
